### PR TITLE
Update aws sdk + s3 improvements

### PR DESCRIPTION
**Pull request by icewind1991**
- [x] requires: https://github.com/nextcloud/3rdparty/pull/51

- move ask sdk from `files_external/3rdparty` to `3rdparty`
- update aws sdk
- adjust files_external s3 and object store s3 to work with the updated sdk
- improve performance for s3 external storage

external storage upload performance comparison: [blackfire](https://blackfire.io/profiles/compare/5a695958-7d3a-4490-aa8c-1fe1b5e1ee8e/graph) #

### CODE REVIEW

**Comment by Xuanwo** [comment]
As we talked in https://github.com/nextcloud/server/pull/3910 , is there any plan to build a new app and move aws sdk into it, so that we can support more object storage backend?

**Comment by icewind1991** [comment]
You can already create an app to add support for an object store backend, simply have the app define the class implementing the backend.

Since the object store config specifies the class to use directly there is no need to do any further "hooking up" of things

**Comment by icewind1991** [comment]
This is ready for review imo

**Comment by nickvergessen** [comment]
hmm couldnt the deps be in the files_externals composer instead of the global one?

**Comment by icewind1991** [comment]
I moved them to the global one since it is also used by the object store backend which is in core

**Comment by MorrisJobke** [comment]
@icewind1991 Could you please fix the updated 3rdparty

**Comment by icewind1991** [comment]
rebased

**Comment by MorrisJobke** [comment]
I resolved the conflict and updated the 3rdparty module.

**Comment by MorrisJobke** [comment]
I get this when I run the installation:

```
{
    "app": "files_external",
    "level": 3,
    "message": {
    	"Exception":"Aws\\S3\\Exception\\S3Exception",
    	"Message":"Error executing \"CreateBucket\" on \"http://abc.127.0.0.1:4567/\"; AWS HTTP error: cURL error 6: Couldn't resolve host name",
    	"Code":0,
    	"Trace":"
    		#0 3rdparty/aws/aws-sdk-php/src/WrappedHttpHandler.php(97): Aws\\WrappedHttpHandler->parseError(Array, Object(GuzzleHttp\\Psr7\\Request), Object(Aws\\Command), Array)
    		#1 3rdparty/guzzlehttp/promises/src/Promise.php(203): Aws\\WrappedHttpHandler->Aws\\{closure}(Array)
    		#2 3rdparty/guzzlehttp/promises/src/Promise.php(174): GuzzleHttp\\Promise\\Promise::callHandler(2, Array, Array)
    		#3 3rdparty/guzzlehttp/promises/src/RejectedPromise.php(40): GuzzleHttp\\Promise\\Promise::GuzzleHttp\\Promise\\{closure}(Array)
    		#4 3rdparty/guzzlehttp/promises/src/TaskQueue.php(47): GuzzleHttp\\Promise\\RejectedPromise::GuzzleHttp\\Promise\\{closure}()
    		#5 3rdparty/guzzlehttp/promises/src/Promise.php(234): GuzzleHttp\\Promise\\TaskQueue->run()
    		#6 3rdparty/guzzlehttp/promises/src/Promise.php(267): GuzzleHttp\\Promise\\Promise->waitIfPending()
    		#7 3rdparty/guzzlehttp/promises/src/Promise.php(225): GuzzleHttp\\Promise\\Promise->invokeWaitList()
    		#8 3rdparty/guzzlehttp/promises/src/Promise.php(267): GuzzleHttp\\Promise\\Promise->waitIfPending()
    		#9 3rdparty/guzzlehttp/promises/src/Promise.php(225): GuzzleHttp\\Promise\\Promise->invokeWaitList()
    		#10 3rdparty/guzzlehttp/promises/src/Promise.php(267): GuzzleHttp\\Promise\\Promise->waitIfPending()
    		#11 3rdparty/guzzlehttp/promises/src/Promise.php(225): GuzzleHttp\\Promise\\Promise->invokeWaitList()
    		#12 3rdparty/guzzlehttp/promises/src/Promise.php(62): GuzzleHttp\\Promise\\Promise->waitIfPending()
    		#13 3rdparty/aws/aws-sdk-php/src/AwsClientTrait.php(59): GuzzleHttp\\Promise\\Promise->wait()
    		#14 3rdparty/aws/aws-sdk-php/src/AwsClientTrait.php(78): Aws\\AwsClient->execute(Object(Aws\\Command))
    		#15 lib/private/Files/ObjectStore/S3ConnectionTrait.php(102): Aws\\AwsClient->__call('createBucket', Array)
    		#16 lib/private/Files/ObjectStore/S3ObjectTrait.php(63): OC\\Files\\ObjectStore\\S3->getConnection()
    		#17 lib/private/Files/ObjectStore/ObjectStoreStorage.php(367): OC\\Files\\ObjectStore\\S3->writeObject('urn:oid:4', Resource id #778)
    		#18 lib/private/Files/Storage/Wrapper/Wrapper.php(350): OC\\Files\\ObjectStore\\ObjectStoreStorage->touch('files/welcome.t...', 1501588075)
    		#19 lib/private/Files/Storage/Wrapper/Availability.php(375): OC\\Files\\Storage\\Wrapper\\Wrapper->touch('files/welcome.t...', NULL)
    		#20 lib/private/Files/Storage/Wrapper/Wrapper.php(350): OC\\Files\\Storage\\Wrapper\\Availability->touch('files/welcome.t...', NULL)
    		#21 lib/private/Files/View.php(1127): OC\\Files\\Storage\\Wrapper\\Wrapper->touch('files/welcome.t...')
    		#22 lib/private/Files/View.php(550): OC\\Files\\View->basicOperation('touch', '/admin/files/we...', Array, NULL)
    		#23 lib/private/Files/Node/Folder.php(181): OC\\Files\\View->touch('/admin/files/we...')
    		#24 lib/private/legacy/util.php(425): OC\\Files\\Node\\Folder->newFile('welcome.txt')
    		#25 lib/private/legacy/util.php(395): OC_Util::copyr('/srv/projects/s...', Object(OC\\Files\\Node\\Folder))
    		#26 lib/private/User/Session.php(486): OC_Util::copySkeleton('admin', Object(OC\\Files\\Node\\Folder))
    		#27 lib/private/User/Session.php(361): OC\\User\\Session->prepareUserLogin(true)
    		#28 lib/private/User/Session.php(547): OC\\User\\Session->completeLogin(*** sensitive parameters replaced ***)
    		#29 lib/private/User/Session.php(326): OC\\User\\Session->loginWithPassword(*** sensitive parameters replaced ***)
    		#30 lib/private/Setup.php(406): OC\\User\\Session->login(*** sensitive parameters replaced ***)
    		#31 core/Command/Maintenance/Install.php(92): OC\\Setup->install(Array)
    		#32 3rdparty/symfony/console/Command/Command.php(256): OC\\Core\\Command\\Maintenance\\Install->execute(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
    		#33 3rdparty/symfony/console/Application.php(818): Symfony\\Component\\Console\\Command\\Command->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
    		#34 3rdparty/symfony/console/Application.php(186): Symfony\\Component\\Console\\Application->doRunCommand(Object(OC\\Core\\Command\\Maintenance\\Install), Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
    		#35 3rdparty/symfony/console/Application.php(117): Symfony\\Component\\Console\\Application->doRun(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
    		#36 lib/private/Console/Application.php(170): Symfony\\Component\\Console\\Application->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
    		#37 console.php(92): OC\\Console\\Application->run()
    		#38 occ(11): require_once('/srv/projects/s...')
    		#39 {main}",
    	"File":"3rdparty/aws/aws-sdk-php/src/WrappedHttpHandler.php",
    	"Line":192
    },
    "method": "--",
    "remoteAddr": "",
    "reqId": "6L06558ZNHfIqjOZYMY9",
    "time": "2017-08-01T11:47:57+00:00",
    "url": "--",
    "user": "admin",
    "userAgent": "--",
    "version": "13.0.0.2"
}
```

This is the config:

```
'objectstore' =>
		array (
			'class' => 'OC\\Files\\ObjectStore\\S3',
			'arguments' =>
				array (
					'bucket' => 'abc',
					'key' => '123',
					'secret' => 'abc',
					'hostname' => '127.0.0.1',
					'port' => '4567',
					'use_ssl' => false,
					'use_path_style' => 'true',
				),
		),
```

And I'm using fakeS3

**Comment by icewind1991** [comment]
@MorrisJobke should be fixed

**Comment by icewind1991** [comment]
rebased since 3rdparty is merged.

Please review @MorrisJobke @nickvergessen @rullzer

**Comment by icewind1991** [comment]
@MorrisJobke @nickvergessen @rullzer *poke*

**Review by rullzer — APPROVED** [review]
Did not blow up my tiny test instance...

:elephant:

**Review by MorrisJobke — APPROVED** [review]
Let me rebase to check if CI is green.

File: apps/files_external/lib/Lib/Storage/AmazonS3.php [hunk 1]
```diff
@@ -394,7 +456,7 @@ public function touch($path, $mtime = null) {
 		$fileType = $this->filetype($path);
```
Comment (MorrisJobke) [inline apps/files_external/lib/Lib/Storage/AmazonS3.php]: Line 453: phan says:

```
apps/files_external/lib/Lib/Storage/AmazonS3.php:453 PhanUndeclaredClassConstant Reference to constant RFC1123 from undeclared class \Aws\Common\Enum\DateFormat
```

**Review by MorrisJobke — CHANGES_REQUESTED** [review]
And please also update the autoloader then this can be merged.

**Comment by karlitschek** [comment]
Can we backport this?

**Pull request merged**